Estimate fees for joining a stake pool

Anviking opened the pull request:
# Issue Number

WB-32: #1094 #1096 (Found it easier to tackle both at once)

# Overview

Also see commit history, but main points:

- [x] Added API endpoint *GET /stake-pools/{stake-pool}/wallets/{wallet}/fee*
- [x] Added `feeBalance :: CoinSelection -> Word64`. I re-use `selectCoinsForDelegation` together with `feeBalance` to implement the `joinStakePoolFee` handler.
- [x] Integration tests:
    - STAKE_POOLS_ESTIMATE_FEE_01 - fee matches eventual cost
    - STAKE_POOLS_ESTIMATE_FEE_02 - empty wallet cannot estimate fee
    - STAKE_POOLS_ESTIMATE_FEE_03 - can't use byron wallets
    - STAKE_POOLS_ESTIMATE_FEE_04 - invalid pool and wallet ids

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Add estimate join-stake-pool fee endpoint

Anviking committed:
Remove redundant constraint

KtorZ commented on lib/core-integration/src/Test/Integration/Framework/TestData.hs:
```diff
@@ -270,6 +273,14 @@ wildcardsWalletName :: Text
+--
+
+stakePoolId1 :: Text
+stakePoolId1 = "4f8d686a02c6e625b5a59cc9e234f32e5d72987012f9c25c9a6b60ddade197d1"
```
**[change required]**

Use `listStakePools` in integration scenarios to find a stake pool id.

**Why?**

This increases the area of "things-that-have-to-be-changed" every time we make a change in the genesis.yaml. Avoid hard-coded ids as much as possible, especially when they can be easily retrieved by other means.

Anviking replied:
This was intended as a _not-necessarily_ existing pool id… I could change the id and rename it to `nonExistingPoolId`…

Otherwise, fully agree on the "things-that-have-to-be-changed" argument of course.

KtorZ replied:
Ah! If it's a non existing pool id, then, a better name would do!

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -845,6 +847,27 @@ estimatePaymentFee ctx wid recipients = do
+        selectCoinsForDelegation @ctx @s @t @k ctx wid
+    let bp = blockchainParameters wal
+    let opts = feeOpts tl (computeCertFee 1) (bp ^. #getFeePolicy)
+    return $ estimateFee opts sel
```
**[change required]**

Use instead `inputBalance`, `changeBalance` and `outputBalance` to compute the _actual fee_ on a given selection.

**Why?**

There's no need to re-estimate fees whereas they can really be known exactly. This has little importance for Jörmungandr but is _more correct_ and will matter for the Haskell nodes.

KtorZ commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -404,6 +414,79 @@ spec = do
+        w <- fixtureWalletWith ctx [fee - 1]
+        joinStakePoolFee ctx (p ^. #id) w >>= flip verify
+            [ expectResponseCode HTTP.status403
+            , expectErrorMessage "I'm unable to select enough coins to pay for a"
```
**[change required]**

Use `errMsg403DelegationFee` from `Test.Integration.Framework.TestData`

**Why?**

Avoid magic constants in code.

KtorZ commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -404,6 +414,79 @@ spec = do
+            w <- fixtureWallet ctx
+            (_, r) <- joinStakePoolFee ctx (p ^. #id) w
+            print r
+            return $ either (error . show) (getQuantity . amount) r
```
**[suggestion]**

Re-write as: 

```hs
getFromResponse amount <$> joinStakePoolFee ctx (p ^. #id) w
```
**Why?**

The intent is clearer and the overall code is more concise.

KtorZ commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -404,6 +414,79 @@ spec = do
+        -- Verify the wallet is now delegating
+        request @ApiWallet ctx (getWalletEp w) Default Empty >>= flip verify
+            [ expectFieldEqual delegation (Delegating (p ^. #id))
+            ]
```
**[question]** 
Why joining a stake pool after looking for the fee? Shouldn't this maybe check that the actual fees are somewhat closed to the estimated one? (they should be exact for Jörmungandr though..).

Anviking replied:
This was testing that a `fixtureWalletWith [fee]` can join **for sure**. It found the [edge-case bug](https://input-output-rnd.slack.com/archives/GBT05825V/p1575449724031000) in the coin-selection when the fee-coefficient was enabled.

But you have a point, it is not relevant to fee estimation; I have removed removed it from this pr.

Anviking committed:
Add feeBalance :: CoinSelection -> Word64

Anviking committed:
Add /stake-pools/{pid}/wallets/{wid}/fee to swagger.yaml

Anviking committed:
Fix incorrect swagger schema for tx fee estimation

Anviking commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -493,6 +503,52 @@ spec = do
+        \empty wallet cannot estimate fee" $ \ctx -> do
+        (_, p:_) <- eventually $
+            unsafeRequest @[ApiStakePool] ctx listStakePoolsEp Empty
+        w <- emptyWallet ctx
```
We are now not testing edge-cases. I think @KtorZ [suggested](https://input-output-rnd.slack.com/archives/GBT05825V/p1575459104040400?thread_ts=1575309301.012300&cid=GBT05825V) that there are existing unit tests for coin selection that should be relevant and better suited for edge-cases. (haven't looked for them myself yet)

jonathanknowles commented on lib/core-integration/src/Test/Integration/Framework/TestData.hs:
```diff
@@ -270,6 +274,23 @@ wildcardsWalletName :: Text
+--
+
+nonExistingStakePool1 :: Text
+nonExistingStakePool1 = "008d686a02c6e625b5a59cc9e234f32e5d72987012f9c25c9a6b60ddade197d1"
```
```suggestion
nonExistingStakePool1 =
    "008d686a02c6e625b5a59cc9e234f32e5d72987012f9c25c9a6b60ddade197d1"
```

paweljakubas replied:
@Anviking  why not to use something like this : 
```
let poolIdAbsent = PoolId $ BS.pack $ replicate 32 0
```
it is already used in `"STAKE_POOLS_JOIN_01 - Cannot join non-existant stakepool"`
then you can invoke `toText` on it ...

jonathanknowles commented on lib/core-integration/src/Test/Integration/Framework/TestData.hs:
```diff
@@ -270,6 +274,23 @@ wildcardsWalletName :: Text
+        [ ("64 chars non-hex", replicate 64 'ś')
+        , ("63 chars hex", replicate 63 '1')
+        , ("65 chars hex", replicate 65 '1')
+        , ("64 chars hex", replicate 64 '1')
```
Question: Is this one (64 hexademical characters) actually invalid? If so, why is it invalid?

(Apologies if I have missed something obvious!)

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api.hs:
```diff
@@ -282,6 +283,15 @@ type JoinStakePool t = "stake-pools"
     :> PutAcccepted '[JSON] (ApiTransaction t)
 
+
+-- | https://input-output-hk.github.io/cardano-wallet/api/#operation/joinStakePool
```
I suspect this one should be `joinStakePoolFee`. So:
```suggestion
-- | https://input-output-hk.github.io/cardano-wallet/api/#operation/joinStakePoolFee
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -836,6 +837,27 @@ joinStakePool ctx spl (ApiT pid) (ApiT wid) (ApiWalletPassphrase (ApiT pwd)) = d
+    -> ApiT PoolId
+    -> ApiT WalletId
+    -> Handler ApiFee
+joinStakePoolFee ctx _spl (ApiT _pid) (ApiT wid) = do
```
At first glance, it seems as though the following arguments can be completely omitted:

1. `_spl :: StakePoolLayer IO`
2. `_pid :: PoolId`

The first argument can be omitted by changing the definition of `stakePools` to:
```hs
stakePools ctx spl =
    listPools spl
    :<|> joinStakePool ctx spl
    :<|> joinStakePoolFee ctx
    :<|> quitStakePool ctx
```

The second argument leads me to another question:

Will the fee calculation always be independent of the pool ID? If the answer is yes, then is there any need for the endpoint to include the pool ID?

KtorZ replied:
I've indeed removed the pool id from the endpoint path and from the handler arguments.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs:
```diff
@@ -92,6 +93,9 @@ outputBalance = foldl' addTxOut 0 . outputs
 changeBalance = foldl' addCoin 0 . change
 
+feeBalance :: CoinSelection -> Word64
+feeBalance sel = inputBalance sel - outputBalance sel - changeBalance sel
```
:+1:

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -493,6 +503,52 @@ spec = do
+        r <- joinStakePool ctx (p ^. #id) (w, fixturePassphrase)
+        verify r
+            [ expectFieldEqual amount fee
+            ]
```
**Question**: Is this expectation valid?

I was under the impression that the precision of fee estimation cannot be guaranteed, because of randomization that occurs during coin selection. See: https://github.com/input-output-hk/cardano-wallet/issues/1094#issuecomment-559773456

If there is randomization going on under the hood, then will this occasionally trigger a test failure?

KtorZ replied:
This is true only by "accident". The wallet test UTxO are all very homogeneous (10 coins of the same amount). So, this compensate the random aspect of it. 

Yet, I agree that checking this in a scenario is misleading and somewhat wrong.

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -1311,6 +1316,20 @@ paths:
+
+        Estimate fee for joining a stake pool.
+      parameters:
+        - *parametersStakePoolId
```
Why is this parameter required?

See https://github.com/input-output-hk/cardano-wallet/pull/1116/files#diff-610bd6b69383f967d3fad205d6c6d86aR854

jonathanknowles reviewed (commented):
Hi @Anviking

Looks mostly good to me! Though I have a few questions (see comments).

My main question:

Why is the `stakePoolId` argument required by the `joinStakePoolFee` endpoint? It seems as though the fee calculation is independent of the particular stake pool being joined. Are we anticipating a future in which the calculation will depend on the target stake pool?

KtorZ committed:
adjust according to review comments

- remove the pool id parameter from the endpoint
- use fee estimator in the tests
- rename 'joinStakePoolFee' to 'delegationFee'

KtorZ reviewed (approved):
rebased + applied reviewed comments.
